sampling/rate_limited/lib/opentelemetry/sampling/rate_limited/sampler.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Sampling
+    module RateLimited
+      class Sampler
+        def initialize(credits_per_second)
+          @credits_per_second = credits_per_second
+          @balance = 0
+          @last_tick = Time.now.to_i
+          @lock = Mutex.new
+        end
+
+        # @param [String] trace_id
+        # @param [OpenTelemetry::Context] parent_context
+        # @param [Enumerable<Link>] links
+        # @param [String] name
+        # @param [Symbol] kind
+        # @param [Hash<String, Object>] attributes
+        # @return [OpenTelemetry::SDK::Trace::Samplers::Result] The sampling result.
+        def should_sample?(trace_id:, parent_context:, links:, name:, kind:, attributes:)
+          if can_spend?
```

Comment:
Sampling is done per span so if you run out of credits_per_second mid trace you will start dropping spans, and potentially end up with broken traces.

Reply:
Wouldn't this be better solved by instantiating a `ParentBased` sampler with the `RateLimited` sampler as `root`? I'D rather avoid making these distinctions here

Reply:
What do you mean by making the parent based sampler as root?

Wouldn't that make it so the rate limited sampling result is always ignored since the Hierarchical Sampler will always check the other samplers to see if it should override the result?